File: Seminar/src/main/java/org/sopt/Seminar/member/service/MemberService.java
package org.sopt.Seminar.member.service;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.sopt.Seminar.member.domain.Member;
import org.sopt.Seminar.member.domain.SOPT;
import org.sopt.Seminar.member.dto.MemberCreateRequest;
import org.sopt.Seminar.member.dto.MemberGetResponse;
import org.sopt.Seminar.member.dto.MemberProfileUpdateRequest;
import org.sopt.Seminar.member.repository.MemberRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class MemberService {

	private final MemberRepository memberRepository;

	public MemberGetResponse getMemberByIdV1(Long id) {
		return MemberGetResponse.of(memberRepository.findById(id).get());
		// Optional 객체 내부가 비었는데 get()을 하게 된다면 Exception 잠재적 발생 가능.
	}

	public MemberGetResponse getMemberByIdV2(Long id) {
		return MemberGetResponse.of(memberRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("존재하지 않는 회원입니다.")));
	}

	public MemberGetResponse getMemberByIdV3(Long id) {
		return MemberGetResponse.of(memberRepository.findByIdOrThrow(id));
	}

	public MemberGetResponse getMemberByIdV4(Long id) {
		return MemberGetResponse.of(findByIdOrThrow(id));
		// v3와 v4는 메소드 구현부만 Repository, Service Layer인지만 다름.
		// java 8 이후에 인터페이스에도 default method 선언이 가능함.
	}

	public Member findByIdOrThrow(Long id) {
		return memberRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("존재하지 않는 회원입니다."));
	}

	public List<MemberGetResponse> getMembers() {
		List<Member> entityList = memberRepository.findAll();

		return entityList.stream().map(MemberGetResponse::of).toList();
	}

	@Transactional
	public String create(MemberCreateRequest request) {
		Member member = Member.builder()
				.name(request.name())
				.nickname(request.nickname())
				.age(request.age())
				.sopt(request.sopt())
				.build();

		return memberRepository.save(member).getId().toString();
	}

	@Transactional
	public void updateSOPT(Long id, MemberProfileUpdateRequest request) {
		Member member = memberRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("존재하지 않는 회원입니다."));
		member.updateSOPT(new SOPT(request.getGeneration(), request.getPart()));
	}

	@Transactional
	public void deleteMember(Long id) {
		memberRepository.deleteById(id);
	}
}

File: kopring/week2/kotlin-semina/src/main/kotlin/org/sopt/kotlinsemina/domain/member/service/MemberService.kt
package org.sopt.kotlinsemina.domain.member.service

import jakarta.persistence.EntityNotFoundException
import org.sopt.kotlinsemina.domain.member.dto.MemberCreateRequest
import org.sopt.kotlinsemina.domain.member.dto.MemberGetResponse
import org.sopt.kotlinsemina.domain.member.dto.MemberProfileUpdateRequest
import org.sopt.kotlinsemina.domain.member.model.Member
import org.sopt.kotlinsemina.domain.member.model.SOPT
import org.sopt.kotlinsemina.domain.member.repository.MemberRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository,
) {

    @Transactional
    fun create(request: MemberCreateRequest): String {
        val member =
            Member(name = request.name, nickname = request.nickname, age = request.age, sopt = request.soptInfo)

        return memberRepository.save(member).id.toString();
    }

    fun getMemberById(memberId: Long): MemberGetResponse {
        val member = memberRepository.findById(memberId).orElseThrow{
            throw EntityNotFoundException("존재하지 않는 회원입니다.")
        } // java에서는 () -> throw new ~~ 이렇게 하는데 문법적으로 약간 다름.

        return MemberGetResponse.of(member)
    }

    fun getMembers(): List<MemberGetResponse> {
        val memberList = memberRepository.findAll()
        return memberList.map { MemberGetResponse.of(it) }.toList()
        // it은 코틀린에서 사용되는 람다 키워드라고 생각하자!
        // 또 java에서는 memberList.stream().map(~) 이런식으로 했던거 같은데 굉장히 간결..?하다
    }

    @Transactional
    fun updateSOPT(memberId: Long, request: MemberProfileUpdateRequest) {
        val member = memberRepository.findById(memberId).orElseThrow {
            throw EntityNotFoundException("존재하지 않는 회원입니다.")
        }

        member.updateSOPT(SOPT(request.generation, request.part)) // Transactional 안붙이고 왜 더티체킹 안되는지만 확인하다가.. 시간버림
    }

    @Transactional
    fun deleteMember(memberId: Long) {
        memberRepository.deleteById(memberId);
    }
}